apollo-livedata-support/src/main/java/com/apollographql/apollo/livedata/LiveDataApollo.kt

```diff
@@ -0,0 +1,240 @@
+
+import android.arch.lifecycle.LiveData
+import android.support.annotation.MainThread
+import android.support.annotation.NonNull
+import com.apollographql.apollo.ApolloCall
+import com.apollographql.apollo.ApolloPrefetch
+import com.apollographql.apollo.ApolloQueryWatcher
+import com.apollographql.apollo.ApolloSubscriptionCall
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.api.internal.Utils.checkNotNull
+import com.apollographql.apollo.cache.normalized.ApolloStoreOperation
+import com.apollographql.apollo.exception.ApolloException
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * The LiveDataApollo class provides methods for converting ApolloCall, ApolloPrefetch and ApolloWatcher types to LiveData
+ * sources.
+ */
+@Suppress("unused")
+object LiveDataApollo {
+    /**
+     * Converts an [ApolloQueryWatcher] to a LiveData.
+     *
+     * @param watcher the ApolloQueryWatcher to convert.
+     * @param T       the value type.
+     * @return the converted LiveData.
+     * @throws NullPointerException if watcher == null
+     */
+    fun <T> from(@NonNull watcher: ApolloQueryWatcher<T>): LiveData<ApolloLiveDataResponse<T>> {
+        checkNotNull(watcher, "watcher == null")
+        return object : LiveData<ApolloLiveDataResponse<T>>() {
+            var started = AtomicBoolean(false)
+
+            @MainThread
+            override fun setValue(value: ApolloLiveDataResponse<T>?) {
+                started.set(true)
+                super.setValue(value)
+            }
+
+            override fun postValue(value: ApolloLiveDataResponse<T>?) {
+                started.set(true)
+                super.postValue(value)
+            }
+
+            override fun onActive() {
+                super.onActive()
+                if (started.compareAndSet(false, true)) {
+                    watcher.enqueueAndWatch(object : ApolloCall.Callback<T>() {
+                        override fun onResponse(response: Response<T>) {
+                            postValue(ApolloLiveDataResponse.of { response })
+                        }
+
+                        override fun onFailure(e: ApolloException) {
+                            postValue(ApolloLiveDataResponse.Failure(e))
+                        }
+                    })
+                }
+            }
+        }
+    }
+
+    /**
+     * Converts an [ApolloCall] to a [LiveData]. The number of emissions this LiveData will have is based
+     * on the [com.apollographql.apollo.fetcher.ResponseFetcher] used with the call.
+     *
+     * @param call the ApolloCall to convert.
+     * @param T    the value type.
+     * @return the converted LiveData.
+     * @throws NullPointerException if originalCall == null
+     */
+    fun <T> from(@NonNull call: ApolloCall<T>): LiveData<ApolloLiveDataResponse<T>> {
+        checkNotNull(call, "call == null")
+        return object : LiveData<ApolloLiveDataResponse<T>>() {
+            var started = AtomicBoolean(false)
```

Comment:
The `onActive()` method is called when the LiveData object has an active observer.

LiveData.java
```java
void activeStateChanged(boolean newActive) {
            if (newActive == mActive) {
                return;
            }
            // immediately set active state, so we'd never dispatch anything to inactive
            // owner
            mActive = newActive;
            boolean wasInactive = LiveData.this.mActiveCount == 0;
            LiveData.this.mActiveCount += mActive ? 1 : -1;
            if (wasInactive && mActive) {
                onActive();
            }
            if (LiveData.this.mActiveCount == 0 && !mActive) {
                onInactive();
            }
            if (mActive) {
                dispatchingValue(this);
            }
        }
```

So I think it needs to a flag for prevent to call `ApolloCall` whenever observers are changed.
This [SingleLiveEvent](https://github.com/googlesamples/android-architecture/blob/dev-todo-mvvm-live/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/SingleLiveEvent.java) is an example of a single emitter.

Reply:
it is always called on the main thread. and if you happen to set any value, observers will still not receive it so you don't need to avoid setting value when that flag goes off.